Automattic/wp-calypso
# Make sidebar-notice JITM template dismissable

Sidebar-notice JITMs generated on the server were not dismissable - they weren't rendered as dismissed and recording their dismissal was not done correctly.

They weren't rendered as dismissed as the JITM API response includes `isDismissible` to indicate
that a JITM ought to be rendered as dismissable but the SidebarBannerTemplate was using `dismissPreferenceName` instead, which isn't part of the JITM REST API.

The rendering issue was addressed by using isDismissable to generate a dismissPreferenceName and passing that to the generated Banner

Recording their dismissal worked on a calypso level - it is implemented in `DismissableCard` which persists to the `calypso_preferences` user attrib, this meant that the user wouldn't see the same message again . The JITM generating server code doesn't know about `calypso_preferences` though, it stores dismissals differently. This meant that the server JITM generator code was always returning the same top-ranked JITM and the `DismissableCard` was refusing to render itself due to being dismissed. 

The dismissal recording issue was addressed by passing onDismiss from UpsellNudge to Banner, rather than onDismissClick and meant that the dismissal was recorded in two places - once in `calypso_preferences` via [`DismissableCard`](https://github.com/Automattic/wp-calypso/blob/trunk/client/blocks/dismissible-card/index.jsx), which records dismissals based upon the JITM ID and once in the [Jetpack option `hide_jitm`](https://github.com/Automattic/jetpack/blob/6510767762d7415cfc0bd1fbecd5fed1b995d9c7/projects/packages/jitm/src/class-jitm.php#L274) via the [jitm calypso block](https://github.com/Automattic/wp-calypso/blob/75e663a3d0133b311a83267ccf735171a1935dec/client/blocks/jitm/index.jsx#L73) which records based upon the JITM feature_class. The `calypso_preferences` dismissal recording was then removed by passing the `dismissTemporary` property around.



https://user-images.githubusercontent.com/93301/141107266-6c2771dc-f6ca-4064-9803-7252b385c3de.mp4


#### Changes proposed in this Pull Request

 * Change SidebarBannerTemplate to:
   * Expect `onDismiss` rather than `onDismissClick`, [as this is what is passed to it](https://github.com/Automattic/wp-calypso/blob/75e663a3d0133b311a83267ccf735171a1935dec/client/blocks/jitm/index.jsx#L19)
   * Expect tracksDismissName to be passed in as part of the tracks variable rather than directly, and give it a default value
   * Generate `dismissPreferenceName` if the message `isDismissable` rather than expecting it to be passed in     
 * Change UpsellNudge to pass onDismiss to the Banner it creates, rather than on onDismissClick 

#### Testing instructions

1. Make your site eligible for two server-generated JITMs with the same feature_class and one in a different feature class
   * The easiest way to do this is to open `jitm-engine.php` in your wpcom sandbox and change the eligibility rules & feature_classes for messages that are defined using `with_template( 'sidebar-banner' );`
 2. Load up Calypso using one of the links below
 3. Note that the message appears without a dismiss button like in 'before' below
 4. Change the messages in `jitm-engine.php` in your wpcom sandbox so that they are dismissable (remove a line saying `is_dismissable( false )`)
 5. Add a tracks event to the messages too, something like `with_tracks( 'dismiss', 'calypso_yourname_test', [ 'cta_name' => 'foo-bar-yourname' ] );`
 5. Reload calypso, the same JITM should now appear with a X icon to dismiss it like in 'after' below
 6. Press X to dismiss it
 7. The message should disappear immediately
 8. Reload the page, the message should still not appear but one of the messages with a different feature class should
 9. Press X to dismiss it, it should disappear immedately
 10. Reload the page, none of your altered messages should appear.
 11. Check tracks, you should be able to see a 'calypso_yourname_test' event after 5-10 minutes


Before | After
-------|------
 ![image](https://user-images.githubusercontent.com/93301/140935132-457440fc-69c4-4a9a-bb3f-be2831cddad6.png) | ![image](https://user-images.githubusercontent.com/93301/140935161-5c510e5e-f41e-46a0-8bde-6000341fa7dc.png)




Related to #52368
